### PR TITLE
Add ToolchainComponentsMissing error msg

### LINKED ISSUE #2705: installing nightly with missing components suggests doing 'rustup component remove'
When trying to install a toolchain that's missing components you may get an error message, e.g.:
```text
$ rustup toolchain add nightly-2019-07-23
info: syncing channel updates for 'nightly-2019-07-23-x86_64-unknown-linux-gnu'
info: latest update on 2019-07-23, rust version 1.38.0-nightly (e649e9034 2019-07-22)
error: component 'clippy' for target 'x86_64-unknown-linux-gnu' is unavailable for download for channel nightly-2019-07-23
Sometimes not all components are available in any given nightly. If you don't need the component, you can remove it with:

    rustup component remove --toolchain nightly-2019-07-23 --target x86_64-unknown-linux-gnu clippy
```
The suggested action is useless, however:
```
$ rustup component remove --toolchain nightly-2019-07-23 --target x86_64-unknown-linux-gnu clippy
error: toolchain 'nightly-2019-07-23-x86_64-unknown-linux-gnu' is not installed
```
The correct suggestion would be to run the original command again with `--profile minimal`

### PR DESCRIPTION
**Pull request by 0xPoe**
close https://github.com/rust-lang/rustup/issues/2705

### CODE REVIEW

**Comment by 0xPoe** [comment]
I want to add tests, but it seems there is a new update_from_dist in dist tests, I don't know how to mock test `dist:: update_from_dist `, do you have any suggestions?

File: src/errors.rs [hunk 1]
```diff
@@ -491,6 +495,31 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest, toolchain: &
...
 }
 
+fn components_missing_msg(cs: &[Component], manifest: &Manifest, toolchain: &str) -> String {
+    assert!(!cs.is_empty());
+    let mut buf = vec![];
+
+    let _ = writeln!(
+        buf,
+        "component {} is unavailable for download for channel '{}'",
+        &cs[0].description(manifest),
```
Comment (kinnison) [inline src/errors.rs]: Why only report the first missing component?

File: src/errors.rs [hunk 2]
```diff
@@ -299,6 +299,10 @@ error_chain! {
             description("some requested components are unavailable to download")
             display("{}", component_unavailable_msg(&c, &manifest, &toolchain))
         }
+        ToolchainComponentsMissing(c: Vec<Component>, manifest: Manifest,toolchain: String) {
+            description("some requested components are unavailable to download")
```
Comment (kinnison) [inline src/errors.rs]:
```suggestion
            description("at least one of the requested components is unavailable to download")
```

File: src/errors.rs [hunk 3]
```diff
@@ -491,6 +495,31 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest, toolchain: &
...
+    );
+    if toolchain.starts_with("nightly") {
+        let _ = write!(
+            buf,
+            "Sometimes not all components are available in any given nightly. "
+        );
+    }
+    let _ = write!(
+        buf,
+        "If you don't need the component, you can try minimal installation with:\n\n{}",
```
Comment (kinnison) [inline src/errors.rs]:
```suggestion
        "If you don't need the component, you could try a minimal installation with:\n\n{}",
```

**Review by kinnison — CHANGES_REQUESTED** [review]
I'd suggest trying to test this more as an integration test like those in cli-misc or cli-v2.

There are setup options for those which provide a channel which is missing (for example) RLS.  the `default` profile wants the RLS component, so it should be possible to put together a test which triggers this pathway.

File: src/errors.rs (hunk 1 above)
Comment (0xPoe) [reply src/errors.rs]: Fixed.

**Comment by 0xPoe** [comment]
@kinnison Added test.

**Comment by 0xPoe** [comment]
@kinnison ping~

Ready for review.

**Review by kinnison — APPROVED** [review]

**Pull request merged**